### PR TITLE
feat: Support string decimal cast

### LINKED ISSUE #2926: Support exhaustive String to Decimal casting
### What is the problem the feature request solves?

Decimal cast equivalent of : https://github.com/apache/datafusion-comet/issues/326

### Describe the potential solution

Decimal cast equivalent of : https://github.com/apache/datafusion-comet/issues/326

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by coderfender**
## Which issue does this PR close?

Similar issue as #326

Closes #2926 

## What changes are included in this PR?

Decimal Types 
(This is the tricky one)
1. I added two functions (`is_valid_decimal` and `parse_string_to_decimal`). `is_valid_decimal` to parse the input string and return True if the input is a valid decimal. This follows similar pattern of Spark's parsing logic which indeed relies on BigDecimal implementation
2. `parse_string_to_decimal` does the heavy lifting of parsing the input string into a valid 'i128' which is essentially how the value is stored (along with scale and precision information). This method also checks if the scale is too high / low and fails early.
4. Once the above method returns a valid i128 integer, we use the decimal builder and return the decimal array

### CODE REVIEW

File: native/spark-expr/src/conversion_funcs/cast.rs [hunk 1]
```diff
@@ -1976,6 +1975,363 @@ fn do_cast_string_to_int<
...
+        .downcast_ref::<StringArray>()
+        .ok_or_else(|| SparkError::Internal("Expected string array".to_string()))?;
+
+    let mut decimal_builder = Decimal128Builder::with_capacity(string_array.len());
+
+    for i in 0..string_array.len() {
+        if string_array.is_null(i) {
+            decimal_builder.append_null();
+        } else {
+            let str_value = string_array.value(i).trim();
```
Comment (andygrove) [inline native/spark-expr/src/conversion_funcs/cast.rs]: It would be better to avoid `trim` since this could be expensive (creating a new string). It would be more efficient for the parser to ignore whitespace at the end of this string.

File: native/spark-expr/src/conversion_funcs/cast.rs [hunk 2]
```diff
@@ -1976,6 +1975,363 @@ fn do_cast_string_to_int<
...
+        || s.eq_ignore_ascii_case("infinity")
+        || s.eq_ignore_ascii_case("+infinity")
+        || s.eq_ignore_ascii_case("-inf")
+        || s.eq_ignore_ascii_case("-infinity")
+        || s.eq_ignore_ascii_case("nan")
+    {
+        return Ok(None);
+    }
+
+    if !is_valid_decimal_format(s) {
```
Comment (andygrove) [inline native/spark-expr/src/conversion_funcs/cast.rs]: Iterating over the string twice (once for validation and once for parsing) seems expensive

File: native/spark-expr/src/conversion_funcs/cast.rs [hunk 3]
```diff
@@ -1976,6 +1975,363 @@ fn do_cast_string_to_int<
...
+    };
+
+    let negative = mantissa_str.starts_with('-');
+    let mantissa_str = if negative || mantissa_str.starts_with('+') {
+        &mantissa_str[1..]
+    } else {
+        mantissa_str
+    };
+
+    let split_by_dot: Vec<&str> = mantissa_str.split('.').collect();
```
Comment (andygrove) [inline native/spark-expr/src/conversion_funcs/cast.rs]: Calling `split` seems expensive as well.

File: spark/src/test/scala/org/apache/comet/CometCastSuite.scala [hunk 4]
```diff
@@ -684,22 +684,73 @@ class CometCastSuite extends CometTestBase with AdaptiveSparkPlanHelper {
     }
   }
 
-  ignore("cast StringType to DecimalType(10,2)") {
-    // https://github.com/apache/datafusion-comet/issues/325
-    val values = gen.generateStrings(dataSize, numericPattern, 8).toDF("a")
-    castTest(values, DataTypes.createDecimalType(10, 2))
+  test("cast StringType to DecimalType(10,2)") {
+    // TODO fix for Spark 4.0.0
+    assume(!isSpark40Plus)
```
Comment (andygrove) [inline spark/src/test/scala/org/apache/comet/CometCastSuite.scala]: What is the issue with Spark 4? Do we need to file a new issue and reference it here?

**Comment by andygrove** [comment]
Thanks for moving this to a separate PR @coderfender. I left some initial comments. Could you run some benchmarks? I'd like to get a feel for performance.

**Comment by coderfender** [comment]
Sure @andygrove . Let me run some benchmarks on the cast op and update this thread

**Comment by andygrove** [comment]
@coderfender, the functionality looks good overall. There are likely some edge cases we do not support that we need to handle or document. I have only found a few so far, and here is a suggested test to add.

```scala
  test("cast StringType to DecimalType non-ASCII characters") {
    // TODO fix for Spark 4.0.0
    assume(!isSpark40Plus)
    val values = Seq(
      "\uff11\uff12\uff13", // fullwidth digits １２３
      "123\u0000", // null byte at end
      "12\u00003", // null byte in middle
      "\u0000123", // null byte at start
      null).toDF("a")
    Seq(true, false).foreach(k =>
      castTest(values, DataTypes.createDecimalType(10, 2), testAnsi = k))
  }
```

```
!== Correct Answer - 5 ==           == Spark Answer - 5 ==
 struct<a:string,a:decimal(10,2)>   struct<a:string,a:decimal(10,2)>
 [null,null]                        [null,null]
![ 123,123.00]                      [ 123,null]
 [12 3,null]                        [12 3,null]
![123 ,123.00]                      [123 ,null]
![１２３,123.00]                       [１２３,null]
```

I think it is fine if we do not support everything, but we should mark the cast as `Incompatible` and add documentation explaining the current limitations.

**Comment by coderfender** [comment]
Great catch @andygrove . Let me try and see if I can resolve the issue or move forward with these changes and mark it compatible

**Comment by coderfender** [comment]
Pushed commit to not have expensive 'trim' and split by '.' operations. Working on the rather difficult part of merging decimal validation and parsing in a single pass

**Comment by coderfender** [comment]
Addressed double iteration and we now validate and parse in a single iteration

**Comment by coderfender** [comment]
Benchmarks after adding further optimizations 

```
================================================================================================
Running benchmark DecimalType(2,2)
================================================================================================

OpenJDK 64-Bit Server VM 17.0.16+8-LTS on Mac OS X 16.0
Apple M2 Max
Cast function to : DecimalType(2,2) , ansi mode enabled : false:                         Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------
SQL Parquet - Spark Cast expr from STRING to : DECIMAL(2,2) , ansi mode enabled : false            106            130          14          9.4         106.3       1.0X
SQL Parquet - Comet Cast expr from STRING to : DECIMAL(2,2) , ansi mode enabled : false             94            123          36         10.7          93.5       1.1X


================================================================================================
Running benchmark DecimalType(2,2)
================================================================================================

OpenJDK 64-Bit Server VM 17.0.16+8-LTS on Mac OS X 16.0
Apple M2 Max
Cast function to : DecimalType(2,2) , ansi mode enabled : true:                         Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
----------------------------------------------------------------------------------------------------------------------------------------------------------------------
SQL Parquet - Spark Cast expr from STRING to : DECIMAL(2,2) , ansi mode enabled : true             88            109          18         11.3          88.3       1.0X
SQL Parquet - Comet Cast expr from STRING to : DECIMAL(2,2) , ansi mode enabled : true             78             85           9         12.8          78.2       1.1X


================================================================================================
Running benchmark DecimalType(10,2)
================================================================================================

OpenJDK 64-Bit Server VM 17.0.16+8-LTS on Mac OS X 16.0
Apple M2 Max
Cast function to : DecimalType(10,2) , ansi mode enabled : false:                         Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------------------------------------------------------
SQL Parquet - Spark Cast expr from STRING to : DECIMAL(10,2) , ansi mode enabled : false            102            132          22          9.8         101.7       1.0X
SQL Parquet - Comet Cast expr from STRING to : DECIMAL(10,2) , ansi mode enabled : false             86            104          18         11.7          85.8       1.2X


================================================================================================
Running benchmark DecimalType(10,2)
================================================================================================

OpenJDK 64-Bit Server VM 17.0.16+8-LTS on Mac OS X 16.0
Apple M2 Max
Cast function to : DecimalType(10,2) , ansi mode enabled : true:                         Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------
SQL Parquet - Spark Cast expr from STRING to : DECIMAL(10,2) , ansi mode enabled : true            118            150          34          8.5         118.1       1.0X
SQL Parquet - Comet Cast expr from STRING to : DECIMAL(10,2) , ansi mode enabled : true             86            138          36         11.6          86.1       1.4X


================================================================================================
Running benchmark DecimalType(20,2)
================================================================================================

OpenJDK 64-Bit Server VM 17.0.16+8-LTS on Mac OS X 16.0
Apple M2 Max
Cast function to : DecimalType(20,2) , ansi mode enabled : false:                         Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------------------------------------------------------
SQL Parquet - Spark Cast expr from STRING to : DECIMAL(20,2) , ansi mode enabled : false            158            198          40          6.3         157.7       1.0X
SQL Parquet - Comet Cast expr from STRING to : DECIMAL(20,2) , ansi mode enabled : false            146            150           4          6.8         146.2       1.1X


================================================================================================
Running benchmark DecimalType(20,2)
================================================================================================

OpenJDK 64-Bit Server VM 17.0.16+8-LTS on Mac OS X 16.0
Apple M2 Max
Cast function to : DecimalType(20,2) , ansi mode enabled : true:                         Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------
SQL Parquet - Spark Cast expr from STRING to : DECIMAL(20,2) , ansi mode enabled : true            160            210          39          6.3         159.6       1.0X
SQL Parquet - Comet Cast expr from STRING to : DECIMAL(20,2) , ansi mode enabled : true            145            150           6          6.9         145.0       1.1X


================================================================================================
Running benchmark DecimalType(34,10)
================================================================================================

OpenJDK 64-Bit Server VM 17.0.16+8-LTS on Mac OS X 16.0
Apple M2 Max
Cast function to : DecimalType(34,10) , ansi mode enabled : false:                         Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------
SQL Parquet - Spark Cast expr from STRING to : DECIMAL(34,10) , ansi mode enabled : false            350            365          27          2.9         350.0       1.0X
SQL Parquet - Comet Cast expr from STRING to : DECIMAL(34,10) , ansi mode enabled : false            191            203          12          5.2         191.0       1.8X


================================================================================================
Running benchmark DecimalType(34,10)
================================================================================================

OpenJDK 64-Bit Server VM 17.0.16+8-LTS on Mac OS X 16.0
Apple M2 Max
Cast function to : DecimalType(34,10) , ansi mode enabled : true:                         Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------------------------------------------------------
SQL Parquet - Spark Cast expr from STRING to : DECIMAL(34,10) , ansi mode enabled : true            347            354          12          2.9         346.5       1.0X
SQL Parquet - Comet Cast expr from STRING to : DECIMAL(34,10) , ansi mode enabled : true            192            224          38          5.2         191.6       1.8X

```

**Comment by coderfender** [comment]
@andygrove please take a look whenever you get a chance. Thank you

File: spark/src/test/scala/org/apache/comet/CometCastSuite.scala [hunk 5]
```diff
@@ -684,21 +683,88 @@ class CometCastSuite extends CometTestBase with AdaptiveSparkPlanHelper {
     }
   }
 
+//  This is to pass the first `all cast combinations are covered`
   ignore("cast StringType to DecimalType(10,2)") {
```
Comment (andygrove) [inline spark/src/test/scala/org/apache/comet/CometCastSuite.scala:687]: can this test also be enabled now?

Comment (coderfender) [reply spark/src/test/scala/org/apache/comet/CometCastSuite.scala:687]: @andygrove this test cant be enabled given that we are still marking the casts as incompatible . However I added the same tests (with a different name to make sure they pass) . Enabling this test will fail 'all cast combinations are covered'

**Comment by andygrove** [comment]
Thanks @coderfender. This LGTM but could you also update the compatibility guide. It currently states `Does not support inputs ending with 'd' or 'f'. Does not support 'inf'. Does not support ANSI mode. Returns 0.0 instead of null if input contains no digits` and this is not entirely correct now.

With the docs update, I'll be happy to approve.

**Comment by coderfender** [comment]
@andygrove the docs are already updated AFAIK

**Comment by coderfender** [comment]
Thank you @andygrove

File: native/spark-expr/src/conversion_funcs/cast.rs (hunk 2 above)
Comment (coderfender) [reply native/spark-expr/src/conversion_funcs/cast.rs]: Thank you . I fixed the double iteration issue

File: spark/src/test/scala/org/apache/comet/CometCastSuite.scala (hunk 4 above)
Comment (coderfender) [reply spark/src/test/scala/org/apache/comet/CometCastSuite.scala]: https://github.com/apache/datafusion-comet/commit/614fe135694ef166eecf1b3ef9968dd038787815

**Comment by parthchandra** [comment]
Late comment, but did rust_decimal not work for us?

**Comment by coderfender** [comment]
Yeah . I had a chat with @andygrove about this and preferred to keep the cast logic to ourselves to match with Spark rather than relying on a new library

**Comment by andygrove** [comment]
> Late comment, but did rust_decimal not work for us?

I didn't know this at the time, but there are security issues with rust_decimal (this is currently blocking PR builds in the datafusion core project).
